drivers/adc/adc_stm32.c

```diff
@@ -823,6 +845,72 @@ static int set_resolution(const struct device *dev,
 	return 0;
 }
 
+static int set_sequencer(const struct device *dev)
+{
+	const struct adc_stm32_cfg *config = dev->config;
+	struct adc_stm32_data *data = dev->data;
+	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
+
+	uint8_t channel_id;
+	uint8_t channel_index = 0;
+	uint32_t channels_mask = 0;
+
+	/* Iterate over selected channels in bitmask keeping track of:
+	 * - channel_index: ranging from 0 -> ( data->channel_count - 1 )
+	 * - channel_id: ordinal position of channel in data->channels bitmask
+	 */
+	for (uint32_t channels = data->channels; channels;
+		      channels &= ~BIT(channel_id), channel_index++) {
+		channel_id = find_lsb_set(channels) - 1;
+
+		uint32_t channel = __LL_ADC_DECIMAL_NB_TO_CHANNEL(channel_id);
+
+		channels_mask |= channel;
+
+		adc_stm32_setup_channel(dev, channel_id);
+
+#if ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE)
+		if (config->sequencer_type == FULLY_CONFIGURABLE) {
+#if defined(CONFIG_SOC_SERIES_STM32H7X) || defined(CONFIG_SOC_SERIES_STM32U5X)
+			/*
+			 * Each channel in the sequence must be previously enabled in PCSEL.
+			 * This register controls the analog switch integrated in the IO level.
+			 */
+			LL_ADC_SetChannelPreselection(adc, channel);
+#endif /* CONFIG_SOC_SERIES_STM32H7X || CONFIG_SOC_SERIES_STM32U5X */
+			LL_ADC_REG_SetSequencerRanks(adc, table_rank[channel_index], channel);
+			LL_ADC_REG_SetSequencerLength(adc, table_seq_len[channel_index]);
+		}
+#endif /* ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE) */
+	}
+
+#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
```

Comment:
Can we use `IS_ENABLED` here in order to gain readability ?
Same for all `#if ANY_ADC_SEQUENCER_TYPE_IS`

Reply:
You mean by replacing the ANY_ADC_SEQUENCER_TYPE_IS macro entirely? I don't know how to do that with just IS_ENABLED. The problem lies with U5 that has both a fully configurable ADC (ADC1/2) and a not fully configurable one (ADC4). In this case we need to include all the code. For all other series, we include only one of the two blocks of code.

Or do you mean something like this:
```
#if ANY_ADC_SEQUENCER_TYPE_IS(NOT_FULLY_CONFIGURABLE)
#define IS_NOT_FULLY_CONFIGURABLE 1
#else
#define IS_NOT_FULLY_CONFIGURABLE 0
#endif
#if ANY_ADC_SEQUENCER_TYPE_IS(FULLY_CONFIGURABLE)
#define IS_FULLY_CONFIGURABLE 1
#else
#define IS_FULLY_CONFIGURABLE 0
#endif

#if IS_ENABLED(IS_FULLY_CONFIGURABLE)
```
That's feasible, but I don't think it makes things more readable.